crates/burn-router/src/types.rs

```diff
@@ -0,0 +1,406 @@
+            handle: $handle,
+            shape: $shape,
+        });
+        let tensor = $Backend::float_to_device(tensor, $device);
+        let handle = $Backend::float_tensor_handle(tensor);
+        Handle::$Backend(handle)
+    }};
+    ($BackendA:ident, $BackendB:ident, $handle:expr, $device:expr, $shape:expr) => {{
+        // Byte bridge between two backends
+        let tensor = $BackendA::float_tensor(TensorHandle { handle: $handle, shape: $shape });
+        let data = try_read_sync($BackendA::float_into_data(tensor)).expect(
+            "Failed to read tensor data synchronously. This can happen on platforms that don't support blocking futures like WASM."
+        );
+        let tensor = $BackendB::float_from_data(data, $device);
+        let handle = $BackendB::float_tensor_handle(tensor);
+        Handle::$BackendB(handle)
+    }};
+}
+
+macro_rules! multi_backend_match {
+    ($shape:expr, ($handle:expr, $device:expr) : $DefaultBackend:ident, $($OtherBackend:ident),+) => {
+        multi_backend_match! (
+            @step
+            $shape,
+            ($handle, $device);
+            {
+                (Handle::$DefaultBackend(handle), MultiDevice::$DefaultBackend(device)) => bridge!($DefaultBackend, handle, device, $shape),
+                $(
+                    (Handle::$DefaultBackend(handle), MultiDevice::$OtherBackend(device)) => bridge!($DefaultBackend, $OtherBackend, handle, device, $shape),
+                    (Handle::$OtherBackend(handle), MultiDevice::$DefaultBackend(device)) => bridge!($OtherBackend, $DefaultBackend, handle, device, $shape),
+                    (Handle::$OtherBackend(handle), MultiDevice::$OtherBackend(device)) => bridge!($OtherBackend, handle, device, $shape),
+                )+
+            };
+            $($OtherBackend),+
+        )
+    };
+
+    (@step
+        $shape:expr,
+        $pats:tt;
+        { $($arms:tt)* };
+        $BackendA:ident,
+        $($OtherBackend:ident),+
+    ) => {
+        multi_backend_match! (
+            @step
+            $shape,
+            $pats;
+            {
+                $($arms)*
+                $(
+                    (Handle::$BackendA(handle), MultiDevice::$OtherBackend(device)) => bridge!($BackendA, $OtherBackend, handle, device, $shape),
+                    (Handle::$OtherBackend(handle), MultiDevice::$BackendA(device)) => bridge!($OtherBackend, $BackendA, handle, device, $shape),
+                )*
+            };
+            $($OtherBackend),*
+        )
+    };
+
+    (@step
+        $shape:expr,
+        ($handle:expr, $device:expr);
+        { $($arms:tt)* };
+        $($BackendA:ident)?
+    ) => {
+        match ($handle, $device) {
+            $($arms)*
+        }
+    };
+}
+
+// Implement multi-backend types and byte bridge for up to 4 backends
+impl_multi_backend_types!(duo, B1, B2);
+impl_multi_backend_types!(trio, B1, B2, B3);
+impl_multi_backend_types!(quad, B1, B2, B3, B4);
```

Comment:
Still unsure about `duo`, `trio` and `quad` module names but lmk if you have any other preferences